File: leetcode/algorithms/go/0002-add-two-numbers.go
package leetcode

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	var dummy = &ListNode{}
	var mark = dummy
	var carry = 0
	for l1 != nil || l2 != nil {
		if l1 != nil {
			carry += l1.Val
			l1 = l1.Next
		}
		if l2 != nil {
			carry += l2.Val
			l2 = l2.Next
		}
		mark.Next = &ListNode{Val: carry % 10}
		mark = mark.Next
		carry /= 10
	}
	if carry != 0 {
		mark.Next = &ListNode{Val: 1}
	}
	return dummy.Next
}

File: leetcode/algorithms/javascript/0002-add-two-numbers.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function (l1, l2) {
    let dummy = new ListNode(0);
    let mark = dummy;
    let carry = 0;
    while (l1 || l2) {
        if (l1) {
            carry += l1.val;
            l1 = l1.next;
        }
        if (l2) {
            carry += l2.val;
            l2 = l2.next;
        }
        mark.next = new ListNode(carry % 10);
        mark = mark.next;
        carry = Math.floor(carry / 10);
    }
    if (carry) {
        mark.next = new ListNode(1);
    }
    return dummy.next;
};